### PR TITLE
Move to nanobind

### PR DESCRIPTION
**Pull request by CloseChoice**
## Overview

This is just the first step and should serve as a PoC to implement numba functions in C++.

supports #4327

### CODE REVIEW

**Comment by daidahao** [comment]
@CloseChoice 

For the mypy test, we can exclude the pyi file.

```
[tool.mypy]
exclude = ["shap/benchmark/*", "tests/benchmark/*", "shap/_cutils.pyi"]
```

For the `run_tests`, how about changing `from shap._cutils import compute_exp_val` to:
```
from .._cutils import compute_exp_val
```

File: stubs/shap/_cutils.pyi [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import numpy as np
```
Comment (CloseChoice) [inline stubs/shap/_cutils.pyi]: not quite sure if this file is really needed!

**Comment by CloseChoice** [comment]
@daidahao happy if you could take a look. I believe everything is working now and should be ready to merge. I also copied your approach to `cext` to the gpu code, there is a test failure, but that should not be related to this PR. Also building of the GPU code is optional as mentioned in the readme.

File: tests/shap_values_baselines/test_exact_test_tabular_single_output_auto_masker_single_value_baseline.npz
Comment (CloseChoice) [inline tests/shap_values_baselines/test_exact_test_tabular_single_output_auto_masker_single_value_baseline.npz:1]: hmm, if this was changed then we should investigate why. I don't see a reason why this should be different at first glance.

**Comment by daidahao** [comment]
@CloseChoice Thank you for the quick response!

I will take a close look at the changes soon. One suggestion: could we hold off the merge by a few days? Switching the build backend could be major changes to the distribution. I would like to do more testing in the meantime.

Also, have you considered a `universal2` build for macOS and `ARM64` build for Windows? We could add those to the cibuildwheels target.

File: pyproject.toml [hunk 2]
```diff
@@ -98,6 +104,8 @@ test = [
   "selenium",  # needed to test the javascript based plots
 ]
 
+build = ["nanobind>=2.12"]
```
Comment (daidahao) [inline pyproject.toml]: We should remove this line here.

File: pyproject.toml [hunk 3]
```diff
@@ -46,6 +51,7 @@ classifiers = [
 ]
 dynamic = ["version"]
 
+
 [project.optional-dependencies]
```
Comment (daidahao) [inline pyproject.toml:55]: There are many optional dependencies that are test-related and should go into `dependency-groups`, so they are not accessible to users via `pip`. We should move these groups here to `dependency-groups`.

File: pyproject.toml [hunk 4]
```diff
@@ -204,13 +257,21 @@ packages = [
...
+  'shap.models',
 ]
 
-[tool.setuptools_scm]
-version_file = "shap/_version.py"
-# Use "no-local-version" so dev releases are compatibile with PyPI
-local_scheme = "no-local-version"
+metadata.version.provider = "scikit_build_core.metadata.setuptools_scm"
+sdist.include = ["shap/_version.py"]
+
```
Comment (daidahao) [inline pyproject.toml:263]: I would recommend adding:
```
sdist.exclude = ["docs*", "notebooks*"]
```

These two folders are more than 200MB+ in sdist.

File: pyproject.toml [hunk 5]
```diff
@@ -313,13 +379,19 @@ test-extras = ["test-core", "plots"]
...
 [tool.cibuildwheel.linux]
 archs = ["x86_64", "aarch64"]
 
 [tool.cibuildwheel.windows]
+# TODO: Should we support Windows ARM64 too?
 archs = ["AMD64"]
 
 [tool.cibuildwheel.macos]
+# TODO: Should we build universal2 wheels for macOS to reduce wheel count?
 archs = ["x86_64", "arm64"]
```
Comment (daidahao) [inline pyproject.toml:393]: We could build universal2 for both x86_64 and arm64, to reduce the wheel count.

```suggestion
archs = ["universal2"]
```

File: pyproject.toml [hunk 6]
```diff
@@ -313,13 +379,19 @@ test-extras = ["test-core", "plots"]
...
 # skip tests on *-musllinux*" since llvmlite and numba do not provide musllinux wheels
-test-skip = "cp314-macosx_x86_64 *-*linux_{aarch64,ppc64le,s390x} *-macosx_arm64 *-musllinux*"
+test-skip = "cp314-macosx_x86_64 *-*linux_{aarch64} *-macosx_arm64 *-musllinux*"
 
 [tool.cibuildwheel.linux]
 archs = ["x86_64", "aarch64"]
 
 [tool.cibuildwheel.windows]
+# TODO: Should we support Windows ARM64 too?
 archs = ["AMD64"]
```
Comment (daidahao) [inline pyproject.toml:389]: Should we support ARM64 too?

```suggestion
archs = ["AMD64", "ARM64"]
```

File: pyproject.toml [hunk 7]
```diff
@@ -299,12 +360,17 @@ omit = ["shap/benchmark/*"]
...
+# build-frontend = "uv"
+
 # Restrict the set of builds to mirror the wheels available in scipy. See #3028
-# skip *-musllinux_aarch64 since numpy doesn't provid those wheels
-# skip cp38-musllinux_x86_64 since numpy never provided cp38 musllinux wheels
-#  they introduced musllinux in 1.25 when they already dropped cp38
 # skip *t (free-threaded builds) since dependencies like llvmlite don't support it yet
-skip = ["cp314-macosx_x86_64", "cp314t-manylinux_x86_64", "cp314t-macosx_x86_64", "cp314t-win_amd64"]
+skip = [
+  "cp314-macosx_x86_64",
```
Comment (daidahao) [inline pyproject.toml:372]: I believe it is possible to build those wheels now, as numba v0.65.0 supports cp314t. We need to investigate this.

File: pyproject.toml [hunk 8]
```diff
@@ -313,13 +379,19 @@ test-extras = ["test-core", "plots"]
 # skip tests on emulated architectures, as they are very slow
 # skip tests on *-macosx_arm64 , as cibuildwheel does not support tests on arm64 (yet)
 # skip tests on *-musllinux*" since llvmlite and numba do not provide musllinux wheels
-test-skip = "cp314-macosx_x86_64 *-*linux_{aarch64,ppc64le,s390x} *-macosx_arm64 *-musllinux*"
+test-skip = "cp314-macosx_x86_64 *-*linux_{aarch64} *-macosx_arm64 *-musllinux*"
```
Comment (daidahao) [inline pyproject.toml]: I think cibuildwheel now supports testing on macosx_arm64. We need to investigate this.

**Review by daidahao — CHANGES_REQUESTED** [review]
I have left some comments for suggestions. I suggest we address and investigate those before we merge into main branch.

Happy to divide them up if you agree!

File: stubs/shap/_cutils.pyi (hunk 1 above)
Comment (daidahao) [reply stubs/shap/_cutils.pyi]: Let's remove it.

File: shap/cutils/cutils.cpp [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// see https://nanobind.readthedocs.io/en/latest/basics.html#basics and following docs
+#include <nanobind/nanobind.h>
+#include "grey_code_utils.h"
+#include "kernel_explainer_utils.h"
+
+namespace nb = nanobind;
+
+NB_MODULE(_cutils, m)
+{
+    m.def("compute_grey_code_row_values", &compute_grey_code_row_values_1d);
```
Comment (daidahao) [inline shap/cutils/cutils.cpp]: I suggest adding full argument names here, so stub could be properly generated:
```cpp
m.def("compute_grey_code_row_values", &compute_grey_code_row_values_1d, "row_values"_a, "mask"_a, ...);
```

File: shap/cutils/cutils.cpp [hunk 10]
```diff
@@ -0,0 +1,13 @@
+// see https://nanobind.readthedocs.io/en/latest/basics.html#basics and following docs
```
Comment (daidahao) [inline shap/cutils/cutils.cpp:1]: I suggest moving `cutils` folder out of `shap` folder to root, so cpp files are not distributed in binary wheels. They would still be included in sdist.

**Review by daidahao — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 5 above)
Comment (CloseChoice) [reply pyproject.toml:393]: haven't read much about that, are there any downsides with this? We actually reduced the wheel count already through limited ABI.

File: shap/cutils/cutils.cpp (hunk 10 above)
Comment (CloseChoice) [reply shap/cutils/cutils.cpp:1]: Good suggestion, would consider this as optional and work for a future PR though.

**Comment by CloseChoice** [comment]
@daidahao thanks for the review. Happy to split the work however you prefer.

File: pyproject.toml (hunk 6 above)
Comment (CloseChoice) [reply pyproject.toml:389]: I would suggest only adding archs we test against (at minimum in the `build_wheels` job). I would also leave this to a future PR. This PR in itself is a clear improvement and other work depends on it, so I think getting this in and opening an issue for discussion/roadmap of potential arm64 support should do

File: pyproject.toml (hunk 3 above)
Comment (CloseChoice) [reply pyproject.toml:55]: Let's do that for all optional dependencies

File: pyproject.toml (hunk 5 above)
Comment (daidahao) [reply pyproject.toml:393]: @CloseChoice 

`universal2` essentially bundles `x86_64` and `arm64` in one binary wheel. The benefit is we would distribute only one binary for both. `cibuildwheel` would test `x86_64` via Rosetta2 emulation layer and `arm64` natively on macOS arm64 CI.

The downside is that the binary `.so` files would double in size (they are less than 200KB natively).

**Comment by daidahao** [comment]
@CloseChoice 

Following the configuration of `sdist.exclude`, I believe`MANIFEST.in` can be safely deleted. See: https://scikit-build-core.readthedocs.io/en/latest/guide/migration_guide.html

File: tests/shap_values_baselines/test_exact_test_tabular_single_output_auto_masker_single_value_baseline.npz
Comment (CloseChoice) [reply tests/shap_values_baselines/test_exact_test_tabular_single_output_auto_masker_single_value_baseline.npz:1]: Okay, we changed the test, so this is fine!

**Comment by CloseChoice** [comment]
I triggered the build_wheels workflow: https://github.com/shap/shap/actions/runs/24469628116

Once that is done I would suggest to try that out locally on various machines to check if that works correctly. I will do that on windows + ubuntu.

cc @daidahao

File: .github/workflows/build_wheels.yml [hunk 11]
```diff
@@ -37,42 +38,14 @@ jobs:
...
   build_wheels:
-    name: Build ${{ matrix.only }}
-    needs: [generate-wheels-matrix, prepare_data]
+    name: Build ${{ matrix.os }} wheels
+    needs: [prepare_data]
     strategy:
       fail-fast: false
       matrix:
-        include: ${{ fromJson(needs.generate-wheels-matrix.outputs.include) }}
+        os: [ubuntu-latest, ubuntu-24.04-arm, macos-14, windows-latest]
```
Comment (daidahao) [inline .github/workflows/build_wheels.yml]: Should we switch to `macos-latest` and try again?

File: pyproject.toml [hunk 12]
```diff
@@ -299,27 +361,36 @@ omit = ["shap/benchmark/*"]
...
-#  they introduced musllinux in 1.25 when they already dropped cp38
+# skip cp314-macosx_x86_64 due to lack of numba and llvmlite support
 # skip *t (free-threaded builds) since dependencies like llvmlite don't support it yet
-skip = ["cp314-macosx_x86_64", "cp314t-manylinux_x86_64", "cp314t-macosx_x86_64", "cp314t-win_amd64"]
+skip = [
+  "cp314-macosx_x86_64",
+  "cp314t-manylinux_x86_64",
+  "cp314t-macosx_x86_64",
+  "cp314t-win_amd64",
+]
```
Comment (daidahao) [inline pyproject.toml:376]: Seems like cp314t can be built already on many platforms. I recommend we skip cp314 on macos-x86_64 only (not supported by numba).
```py
skip = [
  "cp314-macosx_x86_64",
  "cp314t-macosx_x86_64",
]
```

**Review by daidahao — CHANGES_REQUESTED** [review]
Thanks again @CloseChoice

I've left some comments on fixing the build errors!

Also, several actions are running on Node.js 20 (deprecated soon) and generated warnings. We could update them to newer versions running on Node.js 24:
- actions/checkout@v4 -> v6
- actions/setup-python@v5 -> v6
- actions/upload-artifact@v4 -> v6
- actions/download-artifact@v4 -> v7

I will try out macOS builds once they have passed the tests.

**Comment by daidahao** [comment]
Also worth mentioning that this PR fixes #4075 and #3709 once merged, as it includes #4545.

File: pyproject.toml (hunk 12 above)
Comment (CloseChoice) [reply pyproject.toml:376]: Hmm, I think to postpone this to a later PR. I found the error in the stub generation and therefore just splitted use separate runners for macos on arm and x86_64. See this new pipeline: https://github.com/shap/shap/actions/runs/24473745530

**Comment by CloseChoice** [comment]
> Also worth mentioning that this PR fixes #4075 and #3709 once merged, as it includes #4545.

future note to myself: let's keep future PRs smaller, better to understand and to revert if needed. Your #4545 would have been a great standalone PR in itself.

File: pyproject.toml (hunk 12 above)
Comment (daidahao) [reply pyproject.toml:376]: I think `macos-13` is no longer available. See https://docs.github.com/en/actions/reference/runners/github-hosted-runners

I would recommend either `macos-15-intel` or `macos-latest` (arm-based). For the latter, `cibuildwheel` support cross-compiling and can test x86 via Rosetta2.

**Comment by CloseChoice** [comment]
Here's the test pypi page for our generated wheels: https://test.pypi.org/project/shap/0.51.1.dev106/

Somehow the deployment link does not point to the correct release

**Comment by daidahao** [comment]
> Here's the test pypi page for our generated wheels: https://test.pypi.org/project/shap/0.51.1.dev106/
> 
> Somehow the deployment link does not point to the correct release

Thanks! I will try to do some testing on my local macOS and Ubuntu.

**Comment by daidahao** [comment]
@CloseChoice 

I've ran the test suite locally on macOS 26 arm64:
- Full test suite for py 3.12, 3.13, 3.14: **2 failed**
- Core test suite for py 3.14t (some dependencies like tokenizers fail to compile, blocking a full test suite): **ALL PASSED**

Here is the testing script:
```sh
# !/bin/zsh
set -e

ARM64_WHEEL=shap-0.51.1.dev106-cp312-abi3-macosx_11_0_arm64.whl
INTEL_WHEEL=shap-0.51.1.dev106-cp312-abi3-macosx_10_13_x86_64.whl
PY314T_WHEEL=shap-0.51.1.dev106-cp314-cp314t-macosx_11_0_arm64.whl

PYTHON_VERSIONS=("3.12" "3.13" "3.14" "3.14t")

for python_version in "${PYTHON_VERSIONS[@]}"; do
    echo "Testing with Python $python_version"
    if [[ "$python_version" == "3.14t" ]]; then
        WHEELS=("$PY314T_WHEEL")
        EXTRAS="test-core,plots"
    else
        WHEELS=("$ARM64_WHEEL")
        EXTRAS="test,plots"
    fi
    for wheel in "${WHEELS[@]}"; do
        echo "Installing wheel $wheel for Python $python_version"
        rm -rf .venv
        uv python pin $python_version
        uv sync
        uv pip install "$wheel[$EXTRAS]"
        uv run pytest ../shap/tests
    done
done
```

The two failed tests are both from force plot js and related to screenshot dimension. Interestingly, the GitHub CIs did not surface the fails on `macos-latest` (macOS 15) https://github.com/shap/shap/blob/01a3db3b3a7961c18119ea091b40701b2054ef99/.github/workflows/run_tests.yml#L72-L73

My suspicion is that this might a os-related issue rather than a bug.

**Limitation:** `INTEL_WHEEL` (`x86_64`) has not been tested yet, as I don't have access to x86 macOS. I will try to figure out emulating x86 via Rosetta2.

```
========================================== test session starts ==========================================
platform darwin -- Python 3.12.12, pytest-9.0.3, pluggy-1.6.0
Matplotlib: 3.10.8
Freetype: 2.6.1
rootdir: /Users/engs2902/Documents/code/shap
configfile: pyproject.toml
plugins: cov-7.1.0, mpl-0.19.0
collected 577 items / 1 deselected / 576 selected                                                       

../shap/tests/actions/test_action.py ...                                                          [  0%]
../shap/tests/benchmark/test_imports.py .                                                         [  0%]
../shap/tests/explainers/other/test_maple.py .                                                    [  0%]
../shap/tests/explainers/other/test_ubjson.py ...                                                 [  1%]
../shap/tests/explainers/test_coalition.py ....                                                   [  2%]
../shap/tests/explainers/test_deep.py s.....s..ssssssssssssssssssss                               [  7%]
../shap/tests/explainers/test_exact.py .............                                              [  9%]
../shap/tests/explainers/test_explainer.py .....                                                  [ 10%]
../shap/tests/explainers/test_gpu_tree.py sssssssssssssssssssssssssssssssssssss                   [ 16%]
../shap/tests/explainers/test_gradient.py .s.sss..                                                [ 18%]
../shap/tests/explainers/test_kernel.py .......................                                   [ 22%]
../shap/tests/explainers/test_linear.py .............x.....                                       [ 25%]
../shap/tests/explainers/test_partition.py sss..sss                                               [ 26%]
../shap/tests/explainers/test_permutation.py ..........                                           [ 28%]
../shap/tests/explainers/test_sampling.py ......                                                  [ 29%]
../shap/tests/explainers/test_tree.py ........ss..................s...................ss......... [ 39%]
.......................................................                                           [ 49%]
../shap/tests/maskers/test_composite.py ...............                                           [ 51%]
../shap/tests/maskers/test_custom.py .                                                            [ 52%]
../shap/tests/maskers/test_fixed.py .....                                                         [ 52%]
../shap/tests/maskers/test_fixed_composite.py s.                                                  [ 53%]
../shap/tests/maskers/test_image.py ....                                                          [ 53%]
../shap/tests/maskers/test_masker.py ....                                                         [ 54%]
../shap/tests/maskers/test_masker_with_explainers.py ..............                               [ 57%]
../shap/tests/maskers/test_output_composite.py ..........                                         [ 58%]
../shap/tests/maskers/test_tabular.py ...............                                             [ 61%]
../shap/tests/maskers/test_text.py ssssssssssss                                                   [ 63%]
../shap/tests/models/test_teacher_forcing_logits.py sss                                           [ 64%]
../shap/tests/models/test_text_generation.py s                                                    [ 64%]
../shap/tests/plots/test_bar.py ...........                                                       [ 66%]
../shap/tests/plots/test_beeswarm.py ...............................                              [ 71%]
../shap/tests/plots/test_colors.py .........                                                      [ 73%]
../shap/tests/plots/test_decision.py ......                                                       [ 74%]
../shap/tests/plots/test_dependence.py ...                                                        [ 74%]
../shap/tests/plots/test_dependence_string_features.py .....                                      [ 75%]
../shap/tests/plots/test_force.py ...........                                                     [ 77%]
../shap/tests/plots/test_force_js.py FF                                                           [ 77%]
../shap/tests/plots/test_group_difference.py .                                                    [ 77%]
../shap/tests/plots/test_heatmap.py ..                                                            [ 78%]
../shap/tests/plots/test_image.py ........                                                        [ 79%]
../shap/tests/plots/test_scatter.py .........                                                     [ 81%]
../shap/tests/plots/test_style.py ......                                                          [ 82%]
../shap/tests/plots/test_summary.py .......s.........X......                                      [ 86%]
../shap/tests/plots/test_text.py ..                                                               [ 86%]
../shap/tests/plots/test_utils.py ....                                                            [ 87%]
../shap/tests/plots/test_violin.py .....                                                          [ 88%]
../shap/tests/plots/test_waterfall.py ........                                                    [ 89%]
../shap/tests/test_datasets.py .......................                                            [ 93%]
../shap/tests/test_explanation.py ..............                                                  [ 96%]
../shap/tests/test_install.py s                                                                   [ 96%]
../shap/tests/utils/test_clustering.py ......                                                     [ 97%]
../shap/tests/utils/test_general.py ..............                                                [ 99%]
../shap/tests/utils/test_masked_model.py .                                                        [100%]

=============================================== FAILURES ================================================
_________________________________________ test_force_js_visual __________________________________________

driver = <selenium.webdriver.chrome.webdriver.WebDriver (session="c276078a51a31d4bf1aec0a293428511")>

    @pytest.mark.skipif(platform.system() == "Windows", reason="Selenium force plot tests have different sizes on windows.")
    def test_force_js_visual(driver):
        """Test that force plot renders correctly."""
        # Create directory for baseline images if it doesn't exist
        baseline_dir = os.path.join(os.path.dirname(__file__), "baseline")
        os.makedirs(baseline_dir, exist_ok=True)
    
        # Get sample force plot
        plot = get_sample_force_plot()
    
        # Define baseline path
        baseline_path = os.path.join(baseline_dir, "test_force_js.png")
    
        # Capture screenshot
        screenshot = capture_plot_screenshot(driver, plot)
    
        # If baseline doesn't exist, save current screenshot as baseline
        if not os.path.exists(baseline_path):
            screenshot.save(baseline_path)
            pytest.skip(f"Baseline image created at {baseline_path}")
    
        # Compare with baseline
        baseline = Image.open(baseline_path)
    
        # Ensure same dimensions
>       assert screenshot.size == baseline.size, "Screenshot dimensions don't match baseline"
E       AssertionError: Screenshot dimensions don't match baseline
E       assert (1000, 457) == (1000, 461)
E         
E         At index 1 diff: 457 != 461
E         Use -v to get more diff

../shap/tests/plots/test_force_js.py:148: AssertionError
______________________________________ test_force_array_js_visual _______________________________________

driver = <selenium.webdriver.chrome.webdriver.WebDriver (session="c276078a51a31d4bf1aec0a293428511")>

    @pytest.mark.skipif(platform.system() == "Windows", reason="Selenium force plot tests have different sizes on windows.")
    def test_force_array_js_visual(driver):
        """Test that force array plot renders correctly."""
        # Create directory for baseline images if it doesn't exist
        baseline_dir = os.path.join(os.path.dirname(__file__), "baseline")
        os.makedirs(baseline_dir, exist_ok=True)
    
        # Get sample force array plot
        plot = get_sample_force_array_plot()
    
        # Define baseline path
        baseline_path = os.path.join(baseline_dir, "test_force_array_js.png")
    
        # Capture screenshot
        screenshot = capture_plot_screenshot(driver, plot, wait_time=3)  # Array plot might need more time
    
        # If baseline doesn't exist, save current screenshot as baseline
        if not os.path.exists(baseline_path):
            screenshot.save(baseline_path)
            pytest.skip(f"Baseline image created at {baseline_path}")
    
        # Compare with baseline
        baseline = Image.open(baseline_path)
    
        # Ensure same dimensions
>       assert screenshot.size == baseline.size, "Screenshot dimensions don't match baseline"
E       AssertionError: Screenshot dimensions don't match baseline
E       assert (1000, 457) == (1000, 461)
E         
E         At index 1 diff: 457 != 461
E         Use -v to get more diff

../shap/tests/plots/test_force_js.py:184: AssertionError
======================================== short test summary info ========================================
FAILED ../shap/tests/plots/test_force_js.py::test_force_js_visual - AssertionError: Screenshot dimensions don't match baseline
FAILED ../shap/tests/plots/test_force_js.py::test_force_array_js_visual - AssertionError: Screenshot dimensions don't match baseline
= 2 failed, 479 passed, 93 skipped, 1 deselected, 1 xfailed, 1 xpassed, 112 warnings in 405.14s (0:06:45) =
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

```

**Comment by CloseChoice** [comment]
@daidahao okay, works both on windows and ubuntu. I'll do one final check, rerun pipelines and merge after

**Comment by CloseChoice** [comment]
Nice, it's in. Thanks a lot @daidahao
